File: packages/sdk/operational.py
import requests
import os
import logging

'''
How to use:

ops = Tune(key='API_KEY')

# Log an event
event = {
    'name': 'user_signup',
    'content': 'A new user has signed up.',
    'category': 'user-activity',
    'notify': True,
    'avatar': '🐍'
}

response = ops.ingest_event(event)
'''

class Tune:
    def __init__(self, key=None, debug=False, test=False, version='v1', base_url='https://api.tune'):
        self.key = key
        self.debug = debug
        self.test = test
        self.version = version
        self.base_url = base_url

        if not self.key:
            self.log_error("Missing API key. Pass it to the constructor or set OPERATIONAL_API_KEY environment variable.")

    def log_error(self, message):
        if self.debug and message:
            logging.error(message)

    def request(self, endpoint, data):
        url = f"{self.base_url}/{endpoint}"
        headers = {
            'Content-Type': 'application/json',
            'Authorization': f'Bearer {self.key}',
        }

        try:
            response = requests.post(url, json=data, headers=headers)
            response.raise_for_status()
            return None
        except requests.exceptions.RequestException as e:
            self.log_error(f"HTTP error: {str(e)}")
            return None

    def ingest_event(self, event):
        if not event:
            self.log_error("Missing event data")
            return None

        if self.test:
            event['test'] = True
            self.log_error('Sending test event')

        response = self.request(f"api/{self.version}/ingest", event)
        self.log_error(response)
        return response
      

File: packages/sdk/operational.rs
use std::env;
use serde_json::Value;
use reqwest::blocking::Client;

// Usage Example (in main.rs)
/*
fn main() {
    let ops = Tune::new(Some("API_KEY"), false, false, None, None);

    let event = serde_json::json!({
        "name": "user_signup",
        "content": "A new user signed up",
        "category": "user-activity",
        "notify": true
    });

    if ops.ingest_event(event) {
        println!("Event logged successfully");
    } else {
        println!("Failed to log event");
    }
}
*/


pub struct Tune {
    key: String,
    debug: bool,
    test: bool,
    version: String,
    base_url: String,
    client: reqwest::blocking::Client,
}

impl Tune {
    pub fn new(key: Option<&str>, debug: bool, test: bool, version: Option<&str>, base_url: Option<&str>) -> Self {
        let api_key = key.map(String::from)
            .or_else(|| env::var("OPERATIONAL_API_KEY").ok())
            .unwrap_or_default();

        if debug && api_key.is_empty() {
            eprintln!("Missing API key. Pass it to the constructor.");
        }

        Tune {
            key: api_key,
            debug,
            test,
            version: version.unwrap_or("v1").to_string(),
            base_url: base_url.unwrap_or("https://api.tune").to_string(),
            client: reqwest::blocking::Client::new(),
        }
    }

    fn log_error(&self, message: &str) {
        if self.debug {
            eprintln!("[Tune Error]: {}", message);
        }
    }

    fn request(&self, endpoint: &str, data: &serde_json::Value) -> bool {
        let url = format!("{}/{}", self.base_url, endpoint);

        let response = self.client
            .post(&url)
            .header("Content-Type", "application/json")
            .bearer_auth(&self.key)
            .json(data)
            .send();

        match response {
            Ok(resp) => {
                if resp.status().is_success() {
                    true
                } else {
                    self.log_error(&format!("HTTP error: {}", resp.status()));
                    false
                }
            }
            Err(err) => {
                self.log_error(&format!("Request error: {}", err));
                false
            }
        }
    }

    pub fn ingest_event(&self, mut event: serde_json::Value) -> bool {
        if event.is_null() {
            self.log_error("Missing event data");
            return false;
        }

        if self.test {
            event["test"] = serde_json::Value::Bool(true);
            self.log_error("Sending test event");
        }

        self.request(&format!("api/{}/ingest", self.version), &event)
    }
}